### PR TITLE
Python: Improve ag-ui tests and coverage

### PR DESCRIPTION
**Pull request by moonbox3**
### Motivation and Context

Improve ag-ui tests and coverage

### Description

Improve ag-ui tests and coverage

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [X] The PR follows the [Contribution Guidelines](https://github.com/microsoft/agent-framework/blob/main/CONTRIBUTING.md)
- [X] All unit tests pass, and I have added new tests where possible
- [ ] **Is this a breaking change?** If yes, add "[BREAKING]" prefix to the title of the PR.

### CODE REVIEW

**Comment by markwallace-microsoft** [comment]
<a href="https://github.com/microsoft/agent-framework/blob/08e779932206992a2ad350f43ec6b896e20c412c/README.md">Python Test Coverage</a><details><summary>Python Test Coverage Report • </summary><table><tr><th>File</th><th>Stmts</th><th>Miss</th><th>Cover</th><th>Missing</th></tr><tbody><tr><td><b>TOTAL</b></td><td><b>22544</b></td><td><b>2544</b></td><td><b>88%</b></td><td>&nbsp;</td></tr></tbody></table><i>report-only-changed-files is enabled. No files were changed during this commit :)</i></details>

## Python Unit Test Overview
| Tests | Skipped | Failures | Errors | Time |
| ----- | ------- | -------- | -------- | ------------------ |
| 4811 | 25 :zzz: | 0 :x: | 0 :fire: | 1m 20s :stopwatch: |

File: python/packages/ag-ui/tests/ag_ui/test_ag_ui_client.py [hunk 1]
```diff
@@ -357,6 +357,78 @@ async def mock_post_run(*args: object, **kwargs: Any) -> AsyncGenerator[dict[str
...
+        async for update in client._inner_get_response(messages=messages, stream=True, options={"tools": [my_tool]}):
+            updates.append(update)
+
+        # Find the function_call content - it should have agui_thread_id
+        for update in updates:
+            for content in update.contents:
+                if content.type == "function_call" and content.name == "my_tool":
+                    assert content.additional_properties is not None
+                    assert "agui_thread_id" in content.additional_properties
+                    break
```
Comment (moonbox3) [inline python/packages/ag-ui/tests/ag_ui/test_ag_ui_client.py:432]: The `break` only exits the inner `for content` loop. If no update contains a `function_call` content with `name == "my_tool"`, the assertions never execute and the test passes vacuously. Add a guard to ensure the match was actually found.

```suggestion
        # Find the function_call content - it should have agui_thread_id
        found = False
        for update in updates:
            for content in update.contents:
                if content.type == "function_call" and content.name == "my_tool":
                    assert content.additional_properties is not None
                    assert "agui_thread_id" in content.additional_properties
                    found = True
                    break
        assert found, "Expected to find function_call content for my_tool"
```

File: python/packages/ag-ui/tests/ag_ui/sse_helpers.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+    """
+    text = response_content.decode("utf-8")
+    events: list[dict[str, Any]] = []
+    for line in text.splitlines():
+        if line.startswith("data: "):
+            payload = line[6:]
+            try:
+                events.append(json.loads(payload))
+            except json.JSONDecodeError:
+                continue
```
Comment (moonbox3) [inline python/packages/ag-ui/tests/ag_ui/sse_helpers.py:28]: Silently swallowing `json.JSONDecodeError` means a test could pass with an incomplete event list if the server emits malformed SSE. Consider collecting parse failures so tests can optionally assert no parse errors occurred.

```suggestion
            except json.JSONDecodeError as exc:
                import warnings
                warnings.warn(f"Failed to parse SSE payload: {payload!r}: {exc}", stacklevel=2)
                continue
```

File: python/packages/ag-ui/tests/ag_ui/event_stream.py [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+                assert mid not in ends, f"Duplicate TEXT_MESSAGE_END for message_id={mid}"
+                ends.add(mid)
+
+        unclosed = set(starts.keys()) - ends
+        assert not unclosed, f"Unclosed text messages: {unclosed}"
+
+    def assert_tool_calls_balanced(self) -> None:
+        """Assert every TOOL_CALL_START has a matching TOOL_CALL_END with the same tool_call_id."""
+        starts: dict[str, int] = {}
+        ends: set[str] = set()
```
Comment (moonbox3) [inline python/packages/ag-ui/tests/ag_ui/event_stream.py:125]: `assert_tool_calls_balanced` silently overwrites `starts[tid]` on duplicate TOOL_CALL_STARTs, unlike `assert_text_messages_balanced` which asserts `mid not in starts`. This inconsistency means duplicate TOOL_CALL_START events for the same tool_call_id would go undetected.

```suggestion
    def assert_tool_calls_balanced(self) -> None:
        """Assert every TOOL_CALL_START has a matching TOOL_CALL_END with the same tool_call_id."""
        starts: dict[str, int] = {}
        ends: set[str] = set()
        for i, event in enumerate(self.events):
            t = self._type_str(event)
            if t == "TOOL_CALL_START":
                tid = event.tool_call_id
                assert tid not in starts, f"Duplicate TOOL_CALL_START for tool_call_id={tid}"
                starts[tid] = i
            elif t == "TOOL_CALL_END":
                tid = event.tool_call_id
                assert tid in starts, f"TOOL_CALL_END for unknown tool_call_id={tid}"
                assert tid not in ends, f"Duplicate TOOL_CALL_END for tool_call_id={tid}"
                ends.add(tid)
```

**Review by moonbox3 — COMMENTED** [review]
## Automated Code Review

**Reviewers:** 3 | **Confidence:** 84%

### ✓ Correctness
> This is a well-structured diff adding comprehensive test infrastructure (EventStream helper, SSE parsing, golden tests) and extensive edge-case coverage for message adapters, workflow runs, and HTTP round-trips. No significant correctness bugs were found. The code follows project conventions (async tests without decorators, consistent import patterns). One test has a control-flow issue where assertions may never execute, silently passing without verifying the intended behavior.

### ✓ Security Reliability
> This diff is entirely test infrastructure: new test helpers (EventStream, SSE parsing), golden scenario tests, HTTP round-trip tests, and unit tests for internal helpers. No production code is changed. There are no secrets, injection risks, unsafe deserialization, or resource leaks. The only reliability concern is the silent swallowing of JSON decode errors in the SSE test helper, which could mask real encoding bugs in tests.

### ✓ Test Coverage
> This is a substantial and well-structured test coverage improvement, adding golden event-stream tests, HTTP round-trip tests, multi-turn tests, and extensive unit tests for helper functions. The tests use meaningful assertions via the EventStream helper class and cover a wide range of scenarios including error handling, edge cases, and multi-turn interrupt/resume flows. There are a few minor gaps: the EventStream helper (which underpins all golden tests) has no unit tests of its own, assert_tool_calls_balanced silently ignores duplicate TOOL_CALL_STARTs unlike its text message counterpart, and two fixtures are defined but unused. None of these are blocking.

### Suggestions
- In test_client_tool_sets_additional_properties, the `break` only exits the inner loop, so if no content matches `type == "function_call" and name == "my_tool"`, the test passes vacuously without verifying anything. Consider adding a flag to assert the match was actually found.
- In `sse_helpers.py`, `parse_sse_response` silently swallows `json.JSONDecodeError` with `continue`. Consider logging or collecting these errors so test authors can detect when the server emits malformed SSE events, which could otherwise cause tests to pass with incomplete event streams.
- The `pytest_configure` hooks in `tests/ag_ui/conftest.py` and `tests/ag_ui/golden/conftest.py` that insert directories into `sys.path` are redundant with the `pyproject.toml` `pythonpath` change. Consider removing the manual `sys.path` manipulation to avoid confusion about which mechanism is authoritative.
- The EventStream helper class in event_stream.py is the assertion backbone for all golden tests but has no unit tests of its own. Consider adding tests that verify assertion methods correctly detect violations (e.g., unbalanced messages, wrong ordering) to guard against false-positive golden tests.
- assert_tool_calls_balanced (event_stream.py:117-130) silently overwrites in the starts dict on duplicate TOOL_CALL_STARTs, unlike assert_text_messages_balanced which explicitly checks for duplicates. This inconsistency could mask bugs where duplicate TOOL_CALL_START events are emitted for the same tool_call_id.
- The collect_events and make_app fixtures defined in conftest.py (lines 258-331) do not appear to be used by any test in this diff. Consider removing them or adding tests that use them to avoid dead code in the test infrastructure.
- The golden/conftest.py sys.path setup is redundant — the parent conftest.py already has the same pytest_configure hook, and pyproject.toml now includes the test directory in pythonpath. Consider removing the golden/conftest.py file.

---
*Automated review by moonbox3's agents*

File: python/packages/ag-ui/tests/ag_ui/sse_helpers.py [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        events.append(ns)
+    return EventStream(events)
+
+
+_FIELD_MAP: dict[str, str] = {
+    "messageId": "message_id",
+    "runId": "run_id",
+    "threadId": "thread_id",
+    "toolCallId": "tool_call_id",
+    "toolCallName": "tool_call_name",
```
[the remark this answers is not in the document]
Comment (moonbox3) [reply python/packages/ag-ui/tests/ag_ui/sse_helpers.py:68]: Fixed in `08e77993`.

File: python/packages/ag-ui/tests/ag_ui/sse_helpers.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (moonbox3) [reply python/packages/ag-ui/tests/ag_ui/sse_helpers.py:28]: Fixed in `08e77993`.

File: python/packages/ag-ui/tests/ag_ui/event_stream.py [hunk 5]
```diff
@@ -0,0 +1,173 @@
...
+            if t == "TOOL_CALL_START":
+                tid = event.tool_call_id
+                starts[tid] = i
+            elif t == "TOOL_CALL_END":
+                tid = event.tool_call_id
+                assert tid in starts, f"TOOL_CALL_END for unknown tool_call_id={tid}"
+                ends.add(tid)
+
+        unclosed = set(starts.keys()) - ends
+        assert not unclosed, f"Unclosed tool calls: {unclosed}"
```
[the remark this answers is not in the document]
Comment (moonbox3) [reply python/packages/ag-ui/tests/ag_ui/event_stream.py:139]: Fixed in `08e77993`.

File: python/packages/ag-ui/tests/ag_ui/test_workflow_run.py [hunk 6]
```diff
@@ -677,3 +696,734 @@ async def _stream():
...
+        assert _coerce_content(42) is None
+
+    def test_auto_function_approval_response_type_attempted(self):
+        """Dict with approved+id+function_call triggers auto-type path (line 180-181)."""
+        # The function injects type="function_approval_response" into a copy,
+        # but Content.from_dict may fail for complex nested types - returns None.
+        value = {
+            "approved": True,
+            "id": "a1",
+            "function_call": {"call_id": "c1", "name": "fn", "arguments": "{}"},
```
[the remark this answers is not in the document]
Comment (moonbox3) [reply python/packages/ag-ui/tests/ag_ui/test_workflow_run.py:805]: Fixed in `08e77993`.

File: python/packages/ag-ui/tests/ag_ui/test_ag_ui_client.py (hunk 1 above)
Comment (moonbox3) [reply python/packages/ag-ui/tests/ag_ui/test_ag_ui_client.py:432]: Fixed in `08e77993`.

File: python/packages/ag-ui/tests/ag_ui/sse_helpers.py (hunk 2 above)
Comment (moonbox3) [reply python/packages/ag-ui/tests/ag_ui/sse_helpers.py:28]: Fixed in `08e77993`.

File: python/packages/ag-ui/tests/ag_ui/event_stream.py (hunk 3 above)
Comment (moonbox3) [reply python/packages/ag-ui/tests/ag_ui/event_stream.py:125]: Fixed in `08e77993`.

**Review by dmytrostruk — APPROVED** [review]

**Review by giles17 — APPROVED** [review]
